src/frontend/src/optimizer/plan_node/generic/cte_ref.rs

```diff
@@ -0,0 +1,90 @@
+// Copyright 2024 RisingWave Labs
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::hash::Hash;
+
+use itertools::Itertools;
+use pretty_xmlish::StrAssocArr;
+use risingwave_common::catalog::Schema;
+
+use super::{impl_distill_unit_from_fields, GenericPlanNode, GenericPlanRef};
+use crate::binder::ShareId;
+use crate::optimizer::property::FunctionalDependencySet;
+use crate::{optimizer, OptimizerContextRef};
+
+#[derive(Clone, Debug)]
+pub struct CteRef<PlanRef> {
+    share_id: ShareId,
+    base: PlanRef,
+}
+
+impl<PlanRef> PartialEq for CteRef<PlanRef> {
+    fn eq(&self, other: &Self) -> bool {
+        self.share_id == other.share_id
+    }
+}
+
+impl<PlanRef> Eq for CteRef<PlanRef> {}
+
+impl<PlanRef> Hash for CteRef<PlanRef> {
+    fn hash<H: std::hash::Hasher>(&self, state: &mut H) {
+        self.share_id.hash(state);
+    }
+}
+
+impl<PlanRef> CteRef<PlanRef> {
+    pub fn new(share_id: ShareId, base: PlanRef) -> Self {
+        Self { share_id, base }
+    }
+}
+
+impl<PlanRef: GenericPlanRef> CteRef<PlanRef> {
+    pub fn get_cte_ref(&self) -> Option<optimizer::plan_node::PlanRef> {
```

Comment:
this is to serve as the "circle reference" purpose back to the constructed `RecursiveUnion`, but I'm not pretty sure if this is okay especially for subsequent `to_batch` and `to_stream`.

cc @TennyZhuang @xiangjinwu.